File: 120-triangle/triangle.java
class Solution {
    public int minimumTotal(List<List<Integer>> triangle) {
        int [][] dp = new int[triangle.size()+1][triangle.size()+1];
        for(int [] arr: dp){
            Arrays.fill(arr,Integer.MIN_VALUE);
        }
        return fun(0,0,triangle,dp);
    }
    public int fun(int level,int j,List<List<Integer>> triangle,int [][] dp){
        if(level == triangle.size()-1){
            return triangle.get(level).get(j);
        }
        if(dp[level][j] != Integer.MIN_VALUE){
            return dp[level][j];
        }
        int min = Integer.MAX_VALUE;
        for(int i = j;i<j+2;i++){
            min = Math.min(fun(level+1,i,triangle,dp),min);
        }
        dp[level][j] = triangle.get(level).get(j) + min;
        return dp[level][j];
    }
}

File: 120-triangle/triangle.py
class Solution:
    def minimumTotal(self, triangle: List[List[int]]) -> int:
        dp = [[0] * (row + 1) for row in range(len(triangle) + 1)]
        for row in range(len(triangle)-1,-1,-1):
            for j in range(len(triangle[row])):
                dp[row][j] = triangle[row][j] + min(dp[row+1][j], dp[row+1][j+1])
        return dp[0][0]

        # def fun(row, index):
        #     if row == len(triangle):
        #         return 0
        #     return triangle[row][index] + min(fun(row+1,index), fun(row+1,index+1))

        # return fun(0,0)